content/blog/2022/ci-cd-improvements/index.md

```diff
@@ -0,0 +1,104 @@
+- We should **use staging hubs as "canary" deployments** and not continue upgrading production hubs if the staging hub fails.
+
+## An overview of our changes
+
+To accomplish this, we needed to identify which hub required an upgrade based on file additions/modifications.
+This took a lot of discussion and iteration on design, and so we share it below in the hopes that it is helpful to others!
+
+### Improvements to our code and structure
+
+We made a few major changes to [the infrastructure repository](https://github.com/2i2c-org/infrastructure) to facilitate the deployment logic described above.
+Here are the major changes we implemented:
+
+- We separated out each hub's configuration into its own file, or set of files. For example, [here is 2i2c's `staging` hub configuration](https://github.com/2i2c-org/infrastructure/blob/master/config/clusters/2i2c/staging.values.yaml).
+- We created a separate `cluster.yaml` file that holds the canonical list of hubs deployed to that cluster and the configuration file(s) associated with each one. For example, [here is 2i2c's GKE cluster configuration](https://github.com/2i2c-org/infrastructure/blob/master/config/clusters/2i2c/cluster.yaml).
+- We updated [our deployer module](https://github.com/2i2c-org/infrastructure/tree/master/deployer) to do the following things:
+  - Inspect the list of files modified in a Pull Request.
+  - From this list, calculate the name of a hub that required an upgrade, and the name of its respective cluster.
+  - Trigger a GitHub action that would deploy changes in-parallel for each cluster/hub pair.
+
+In addition to these structural and code changes, we also developed new GitHub Actions that control the entire process.
+
+### A GitHub Actions workflow for upgrading our JupyterHubs
+
+We defined a new GitHub Action that carries out the logic described above.
+These are all defined in [this `deploy-hubs.yaml` configuration file](https://github.com/2i2c-org/infrastructure/blob/master/.github/workflows/deploy-hubs.yaml).
+Here are the major steps in this action, and what each does:
+
+1. `generate-jobs`: Generate a list of clusters/hubs that must be upgraded, given the files that are changed in a Pull Request.
+   - Evaluate an input list of added/modified files in a PR
+   - Decide if the added/modified files warrant an upgrade of a hub
+   - Generate a list of hubs and clusters that require upgrades, and some extra details:
+     - Does the support chart that is deployed to the cluster also need an upgrade?
+     - Does a staging hub on this cluster require an upgrade?
+
+   This produced two outputs to be used in subsequent steps:
+
+   - A **human-readable table** including information on _why_ a given deployment requires an upgrade (using the excellent [Rich library](https://github.com/Textualize/rich)).
+   - A **JSON output** that can be interpreted by GitHub Actions as a set of matrix jobs to run.
+
+   % TODO: Can we add an image of this nice table in the post?
```

Comment:
I took a couple of partial screenshots - it's quite a long table when everything needs upgrading, so I wasn't able to get everything in without making the font tiny

<img width="840" alt="Screenshot 2022-04-25 at 13 25 32" src="https://user-images.githubusercontent.com/44771837/165088695-cca1750f-9323-4e56-9a79-c3cd5997f9ef.png">

<img width="597" alt="Screenshot 2022-04-25 at 13 25 42" src="https://user-images.githubusercontent.com/44771837/165088688-1577a4f2-14b9-4029-a312-c1adcac6ac7a.png">